staticcheck/lint.go

```diff
@@ -117,6 +117,41 @@ func checkValidHostPort(arg int) CallCheck {
 	}
 }
 
+func checkNonOverlappingDstSrc(dstArg, srcArg int) CallCheck {
+	return func(call *Call) {
+		dst := call.Args[dstArg]
+		src := call.Args[srcArg]
+		_, dstConst := irutil.Flatten(dst.Value.Value).(*ir.Const)
+		_, srcConst := irutil.Flatten(src.Value.Value).(*ir.Const)
+		if dstConst || srcConst {
+			// one of the arguments is nil, therefore overlap is not possible
+			return
+		}
+		if dst.Value == src.Value {
+			// simple case of f(b, b)
+			dst.Invalid("overlapping dst and src")
+			return
+		}
+		dstSlice, ok := irutil.Flatten(dst.Value.Value).(*ir.Slice)
```

Comment:
`ir.Slice` is the IR instruction for slicing. Not all slices come into existence by something being sliced. For example, analyzing this code panics:

```
func foo(a, b []byte) {
	hex.Encode(a, b)
}
```

As does

```
func foo(n int) {
	a := make([]byte, n)
	b := make([]byte, n)
	hex.Encode(a, b)
}
```